pghoard/rohmu/object_storage/sftp.py

```diff
@@ -0,0 +1,209 @@
+"""
+rohmu - sftp object store interface
+
+Copyright (c) 2016 Ohmu Ltd
+See LICENSE for details
+"""
+
+import logging
+import warnings
+import pysftp
+from pysftp import CnOpts
+import os
+import datetime
+from stat import S_ISDIR
+from io import BytesIO, StringIO
+import json
+from .base import BaseTransfer, IterKeyItem, KEY_TYPE_PREFIX, KEY_TYPE_OBJECT
+from ..errors import FileNotFoundFromStorageError, InvalidConfigurationError, StorageError
+
+
+class SFTPTransfer(BaseTransfer):
+    def __init__(self,
+                 server,
+                 port,
+                 username,
+                 password=None,
+                 private_key=None,
+                 prefix=None):
+        super().__init__(prefix=prefix)
+        self.server = server
+        self.port = port
+        self.username = username
+        self.password = password
+        self.private_key = private_key
+
+        if password is None and private_key is None:
+            raise InvalidConfigurationError("Password or private key required")
+
+        # https://bitbucket.org/dundeemt/pysftp/issues/109/hostkeysexception-no-host-keys-found-even
+        warnings.filterwarnings('ignore', '.*hostkeys = None*',)
+        self.cnopts = CnOpts()
+        self.cnopts.hostkeys = None
+
+        self.cnopts.log = False
+        logging.getLogger("paramiko").setLevel(logging.WARNING)
+
+        self.client = pysftp.Connection(self.server,
+                                        port=self.port,
+                                        username=self.username,
+                                        password=self.password,
+                                        private_key=self.private_key,
+                                        cnopts=self.cnopts)
+        self.log.debug("SFTPTransfer initialized")
+
+    def get_contents_to_file(self, key, filepath_to_store_to, *, progress_callback=None):
+        with open(filepath_to_store_to, "wb") as fh:
+            return self.get_contents_to_fileobj(key, fh, progress_callback=progress_callback)
+
+    def get_contents_to_fileobj(self, key, fileobj_to_store_to, *, progress_callback=None):
+        self._get_contents_to_fileobj(key, fileobj_to_store_to, progress_callback)
+        return self.get_metadata_for_key(key)
+
+    def _get_contents_to_fileobj(self, key, fileobj_to_store_to, progress_callback=None):
+        target_path = self.format_key_for_backend(key.strip("/"))
+        self.log.debug("Get file content: %r", target_path)
+
+        try:
+            return self.client.getfo(target_path, fileobj_to_store_to, progress_callback)
```

Comment:
Maybe add comment here mentioning that pysftp conveniently happens to have the exact same signature for progress callback as pghoard. This looks suspicious otherwise.